### PR TITLE
Organize published artifacts

### PR DESCRIPTION
**Pull request by chidozieononiwu**
Organize Generated Artifacts Into Per-Package folders.
Update release stage to work with the new structure.

Depends on https://dev.azure.com/azure-sdk/internal/_git/azure-sdk-build-tools/pullrequest/173
Sample Build https://dev.azure.com/azure-sdk/internal/_build/results?buildId=532871&view=results

### CODE REVIEW

**Comment by weshaggard** [comment]
This has the potential to throw off some releases so I'd like to hold off on taking this change until after the sept releases.

**Comment by weshaggard** [comment]
As part of this effort I'd like to try and eliminate this https://github.com/Azure/azure-sdk-for-net/blob/master/eng/pipelines/templates/steps/archetype-sdk-docs.yml#L23 as well. I'd like to get to a point where we only have the packages (and packages-signed) artifacts and each have a folder per package with the package as well as the docs and any other artifacts nested in the folder with the package. This means it is worth having a look at PublishPipelineArtifact and finding all the artifacts we publish and see out to setup the publishing up to combine those. 

FYI @scbedd

**Comment by chidozieononiwu** [comment]
> As part of this effort I'd like to try and eliminate this https://github.com/Azure/azure-sdk-for-net/blob/master/eng/pipelines/templates/steps/archetype-sdk-docs.yml#L23 as well. I'd like to get to a point where we only have the packages (and packages-signed) artifacts and each have a folder per package with the package as well as the docs and any other artifacts nested in the folder with the package. This means it is worth having a look at PublishPipelineArtifact and finding all the artifacts we publish and see out to setup the publishing up to combine those.
> 
> FYI @scbedd

That is already part of the work here. Here is a[ sample build ](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=520550&view=artifacts&type=publishedArtifacts)showing how the artifacts are published. I also did a template release to test other release steps. But yes we should wait till after September release.

**Review by mitchdenny — APPROVED** [review]
Approved, pending a test run with Azure.Template all the way out.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:34]: It is probably about time we split this to our newer standard of one argument per line.

File: eng/pipelines/templates/stages/archetype-net-release.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-net-release.yml:28]: Why do we need to copy anything? Part of the point of having these be separated in the artifact itself is to avoid the need for copying and partitioning further.

File: eng/pipelines/templates/stages/archetype-net-release.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-net-release.yml:69]: can we delete the stage-artifacts.yml now?

File: eng/pipelines/templates/stages/archetype-net-release.yml
Comment (weshaggard) [reply eng/pipelines/templates/stages/archetype-net-release.yml:28]: If for some reason the sizing steps still need this then we should do a copy using and include of *.nupkg

Comment (chidozieononiwu) [reply eng/pipelines/templates/stages/archetype-net-release.yml:28]: Yes the signing steps requires that we copy out the artifacts because it potentially mutates the artifacts while signing them. Yes we can use the include *.nupkg

File: eng/pipelines/templates/stages/archetype-net-release.yml
Comment (chidozieononiwu) [reply eng/pipelines/templates/stages/archetype-net-release.yml:69]: yes

File: eng/pipelines/templates/stages/archetype-net-release.yml
Comment (weshaggard) [reply eng/pipelines/templates/stages/archetype-net-release.yml:28]: What happens if we simply pass the entire path to the signing step? Also do we need to also exclude the *.symbols.nupkg?

Comment (weshaggard) [reply eng/pipelines/templates/stages/archetype-net-release.yml:28]: It is ok if the packages get signed in place as we are going to be re-publishing them into the packages-signed artifact and we want that to have exactly the same files and structure but just have signed files instead of unsigned. There should be no reason we need to have the unsigned packages as part of the signing job after we do the signing.

Comment (chidozieononiwu) [reply eng/pipelines/templates/stages/archetype-net-release.yml:28]: If this is the case do we need to republish it as `packages-signed` since it will be the same as `packges` except for the presence of docs.

File: eng/common/scripts/copy-docs-to-blobstorage.ps1
Comment (weshaggard) [inline eng/common/scripts/copy-docs-to-blobstorage.ps1:281]: Looks like you lost the releaseTag change here.

File: eng/pipelines/templates/stages/archetype-net-release.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-net-release.yml:118]: We should probably split this line so it is a little easier to read.

File: eng/pipelines/templates/stages/archetype-net-release.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-net-release.yml:179]: Need to update ArtifactLocation here. That was one of @sima-zhu's recent changes. We also need to figure out a way to have that only contain the nupkg and not the symbols package otherwise the release tag step will not work.

Comment (weshaggard) [reply eng/pipelines/templates/stages/archetype-net-release.yml:179]: Please work with @sima-zhu to make sure the master to release-tag stuff still works after your change.

File: eng/pipelines/templates/stages/archetype-sdk-client.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-sdk-client.yml:6]: I assume this will be updated before merging.

File: eng/pipelines/templates/steps/archetype-sdk-docs.yml
Comment (weshaggard) [inline eng/pipelines/templates/steps/archetype-sdk-docs.yml:18]: Can we please turn this into our multiple powershell task invocation since it is only calling one PS script.

File: eng/pipelines/templates/stages/archetype-sdk-client.yml
Comment (chidozieononiwu) [reply eng/pipelines/templates/stages/archetype-sdk-client.yml:6]: Yes, I will update that.

File: eng/pipelines/templates/stages/archetype-net-release.yml
Comment (chidozieononiwu) [reply eng/pipelines/templates/stages/archetype-net-release.yml:179]: I copied the `.nupkg` files to another directory just before calling `RetrieveReleaseTag` in the file.

File: eng/common/scripts/copy-docs-to-blobstorage.ps1
Comment (sima-zhu) [inline eng/common/scripts/copy-docs-to-blobstorage.ps1:269]: Can we utilize VerifyPackage method in artifact-parsing-metadata.ps1? 
First, we can filter the nuget file (not symbol one) in folder "$BinDirectory/publicartifactscopy", then VerifyPackage is able to return us the right package metadata like
```
{ 
    $pkgName: "Azure.Core",
    $pkgVersion: ...
    $ReleaseTag: ...
    $ReadmeContent: ...
}
```

You don't need to parse $pkgName and $version again

Comment (sima-zhu) [reply eng/common/scripts/copy-docs-to-blobstorage.ps1:269]: Also, there is no need to call RetrieveReleaseTag but to return the releaseTag from package metadata.

File: eng/common/scripts/copy-docs-to-blobstorage.ps1
Comment (chidozieononiwu) [reply eng/common/scripts/copy-docs-to-blobstorage.ps1]: Fixed

File: eng/common/scripts/copy-docs-to-blobstorage.ps1
Comment (chidozieononiwu) [reply eng/common/scripts/copy-docs-to-blobstorage.ps1]: Thanks @sima-zhu

File: eng/common/scripts/copy-docs-to-blobstorage.ps1
Comment (weshaggard) [inline eng/common/scripts/copy-docs-to-blobstorage.ps1:278]: Is DocLocation the correct location to pass as the artifactlocation? Would you pass PublicArtifactLocation?

File: eng/common/pipelines/templates/steps/publish-blobs.yml
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/publish-blobs.yml:29]: This doesn't look correct. Why are we passing the SourceVersion as the sha?

File: eng/common/scripts/copy-docs-to-blobstorage.ps1
Comment (weshaggard) [reply eng/common/scripts/copy-docs-to-blobstorage.ps1:278]: Why are we not just using what was already here and calling RetrieveReleaseTag?

File: eng/common/pipelines/templates/steps/publish-blobs.yml
Comment (chidozieononiwu) [reply eng/common/pipelines/templates/steps/publish-blobs.yml:29]: From my understanding the $(Build.SourceVersion) will be the SHA for the current release. Is there a situation where that will not be the case?

Comment (sima-zhu) [reply eng/common/pipelines/templates/steps/publish-blobs.yml:29]: No, releaseTag is not SHA commit. It is like "azure-template_1.0.0-beta.1"

File: eng/common/scripts/copy-docs-to-blobstorage.ps1
Comment (sima-zhu) [reply eng/common/scripts/copy-docs-to-blobstorage.ps1:278]: `Verify-Packages` and `RetrieveReleaseTag` takes in the same artifact location, the `artifactLocation` must include one .nuget file (no symbol one).

`$DocLocation` currently contains two .neget file, so it is not the right one to use.
Once we have $pkgProperties from VerifyPackages, we are able to get all package metadata, so there is no need to call RetrieveReleaseTag.

File: eng/common/pipelines/templates/steps/publish-blobs.yml
Comment (sima-zhu) [reply eng/common/pipelines/templates/steps/publish-blobs.yml:29]: Sorry, I thought it was releaseTag variable. $releaseSha is SHA commit to use in `VerifyPackages` method

File: eng/common/scripts/copy-docs-to-blobstorage.ps1
Comment (weshaggard) [reply eng/common/scripts/copy-docs-to-blobstorage.ps1:278]: VerifyPackages does extra work and needs things like the release sha which I don't think we care about for our purposes. I also think VerifyPackages returns back all the packages even the symbols packages similar to RetrieveReleaseTag and if not then we should make RetrieveReleaseTag work the same way. 

In either case I think the correct fix is to start ignoring the symbols package in the metadata parsing code.

Comment (sima-zhu) [reply eng/common/scripts/copy-docs-to-blobstorage.ps1:278]: I agreed with you for the part of Verify-Packages done much more than RetrieveReleaseTag.
However, it seems extra burden to have packageId and packageVersion to parse out again as we did it in $parseNPMArifact already. Can we utilize the method in languageSetting.ps1?

File: eng/common/scripts/copy-docs-to-blobstorage.ps1
Comment (sima-zhu) [inline eng/common/scripts/copy-docs-to-blobstorage.ps1:267]: Can we print out both counts out to help us understand which one fails the scripts?

File: eng/common/scripts/copy-docs-to-blobstorage.ps1
Comment (sima-zhu) [inline eng/common/scripts/copy-docs-to-blobstorage.ps1:278]: I believe this is good method to use. Thanks for investigating!

**Review by sima-zhu — APPROVED** [review]
Release tag part looks good to me. Please kick off a template pipeline to check.

File: eng/common/scripts/copy-docs-to-blobstorage.ps1
Comment (weshaggard) [inline eng/common/scripts/copy-docs-to-blobstorage.ps1:14]: I assume we can undo this change now.

File: eng/common/pipelines/templates/steps/publish-blobs.yml
Comment (weshaggard) [reply eng/common/pipelines/templates/steps/publish-blobs.yml:29]: Based on the update we should be able to undo this change

File: eng/pipelines/templates/stages/archetype-sdk-client.yml
Comment (weshaggard) [reply eng/pipelines/templates/stages/archetype-sdk-client.yml:6]: Didn't you merge your azure-sdk-build-tools PR?

File: eng/common/scripts/copy-docs-to-blobstorage.ps1
Comment (weshaggard) [inline eng/common/scripts/copy-docs-to-blobstorage.ps1:10]: Just a reminder these changes will need to be made in the tools repo. Soon this work will be in the language repo once you get your next set of refactorings done.

Comment (chidozieononiwu) [reply eng/common/scripts/copy-docs-to-blobstorage.ps1:10]: Yh, I will update that PR

File: eng/pipelines/templates/stages/archetype-sdk-client.yml
Comment (chidozieononiwu) [reply eng/pipelines/templates/stages/archetype-sdk-client.yml:6]: Not Yet.

File: doc/ApiDocGeneration/Generate-Api-Docs.ps1
Comment (weshaggard) [inline doc/ApiDocGeneration/Generate-Api-Docs.ps1:144]: Is there a reason we zipping these files? We could also just put them in a folder that would make this step and the extraction step unnecessary.

Comment (chidozieononiwu) [reply doc/ApiDocGeneration/Generate-Api-Docs.ps1:144]: No particular reason, other that it make downloading from the UI a bit faster because otherwise DevOps first zips it before downloading it.

Comment (weshaggard) [reply doc/ApiDocGeneration/Generate-Api-Docs.ps1:144]: Ok in that case I'd suggest we just avoid this round tripping and just put the files directly in an artifact folder.

Comment (weshaggard) [reply doc/ApiDocGeneration/Generate-Api-Docs.ps1:144]: Once you finish your other work to move the language specific implementation of this into the .NET repo I would still like us to change this from being a zip file to just a folder in the artifacts.

**Comment by sima-zhu** [comment]
@chidozieononiwu I did not see any template pipeline which read ReleaseTag correctly. Did you verify the release tag changes?
image
As you can see, the release tag is empty here, which I expect to see it gets print out.

**Comment by chidozieononiwu** [comment]
> @chidozieononiwu I did not see any template pipeline which read ReleaseTag correctly. Did you verify the release tag changes?
> image
> As you can see, the release tag is empty here, which I expect to see it gets print out.

Thanks for pointing that out. I fixed it in this PR https://github.com/Azure/azure-sdk-tools/pull/1022

**Pull request merged**
